File: dining_philosopher.py
import threading
import time

NUM_PHILOSOPHERS = 5

class Philosopher(threading.Thread):
    def __init__(self, id, left_chopstick, right_chopstick):
        super().__init__()
        self.id = id
        self.left_chopstick = left_chopstick
        self.right_chopstick = right_chopstick

    def run(self):
        while True:
            print(f"Philosopher {self.id} is thinking")
            time.sleep(1)

            print(f"Philosopher {self.id} is hungry")
            self.left_chopstick.acquire()
            self.right_chopstick.acquire()

            print(f"Philosopher {self.id} is eating")
            time.sleep(1)

            self.right_chopstick.release()
            self.left_chopstick.release()

if __name__ == "__main__":
    chopsticks = [threading.Semaphore(1) for _ in range(NUM_PHILOSOPHERS)]
    philosophers = []

    for i in range(NUM_PHILOSOPHERS):
        left_chopstick = chopsticks[i]
        right_chopstick = chopsticks[(i + 1) % NUM_PHILOSOPHERS]
        philosophers.append(Philosopher(i, left_chopstick, right_chopstick))

    for philosopher in philosophers:
        philosopher.start()

    for philosopher in philosophers:
        philosopher.join()

File: dining_philosopher.cpp
#include <iostream>
#include <thread>
#include <chrono>
#include <vector>
#include <semaphore.h>

using namespace std;

const int NUM_PHILOSOPHERS = 5;

sem_t chopsticks[NUM_PHILOSOPHERS];
vector<thread> philosophers;

void philosopher(int id) {
    int left_chopstick = id;
    int right_chopstick = (id + 1) % NUM_PHILOSOPHERS;

    while (true) {
        cout << "Philosopher " << id << " is thinking" << endl;
        this_thread::sleep_for(chrono::milliseconds(1000));

        cout << "Philosopher " << id << " is hungry" << endl;

        sem_wait(&chopsticks[left_chopstick]);

        sem_wait(&chopsticks[right_chopstick]);

        cout << "Philosopher " << id << " is eating" << endl;
        this_thread::sleep_for(chrono::milliseconds(1000));

        sem_post(&chopsticks[left_chopstick]);

        sem_post(&chopsticks[right_chopstick]);
    }
}

int main() {
    for (int i = 0; i < NUM_PHILOSOPHERS; i++) {
        sem_init(&chopsticks[i], 0, 1);
    }

    for (int i = 0; i < NUM_PHILOSOPHERS; i++) {
        philosophers.emplace_back(philosopher, i);
    }

    for (auto& p : philosophers) {
        p.join();
    }

    for (int i = 0; i < NUM_PHILOSOPHERS; i++) {
        sem_destroy(&chopsticks[i]);
    }

    return 0;
}
